### PR TITLE
Feature/gloo spark lightfm wrapper 0.2

### PR DESCRIPTION
**Pull request by zakharova-anastasiia**

### CODE REVIEW

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+        features_np = (
+            features.select(f"{entity}_idx", *feature_names)
+            .toPandas()
+            .to_numpy()
+        )
+        feature_entity_ids = features_np[:, 0]
+        features_columns = features_np[:, 1:]
+        number_of_features = features_columns.shape[1]
+        # Scale down dense features
+        scaler_name = f"{entity}_feat_scaler"
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:418]: Explicit is better than implicit.

def __init__(....)
     self.__item_feat_scaler: Optional[Scaler] = None
     self.__user_feat_scaler: Optional[Scaler] = None


def __get_scaler(entity: str) -> Scaler:
    # create or choose from __item_feat_scaler or __user_feat_scaler
     ...

if self.__feat_scaler is None:
    pass

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        if self.__getattribute__(scaler_name) is None:
+            if not features_columns.size:
+                raise ValueError(f"features for {entity}s from log are absent")
+            self.__setattr__(scaler_name, MinMaxScaler().fit(features_columns))
+
+        if features_columns.size:
+            features_dense = self.__getattribute__(scaler_name).transform(
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:425]: replace __getattribute__

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+            k=self.k_kos_warp,
+            n=self.n_kos_warp,
+            max_sampled=self.max_sampled,
+            random_state=self.random_state,
+        )
+
+    def _convert_features_to_csr(
+        self, entity_ids: DataFrame, features: Optional[DataFrame] = None
+    ) -> Optional[sp.csr_matrix]:
+        """
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:366]: add inputs to the docstring

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+        "item_bias_momentum",
+        "user_embeddings",
+        "user_embedding_gradients",
+        "user_embedding_momentum",
+        "user_biases",
+        "user_bias_gradients",
+        "user_bias_momentum",
+    )
+    losses = ["warp", "bpr", "warp-kos", "logistic"]
+    user_feat_scaler: Optional[MinMaxScaler] = None
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:61]: remove from here and move to __init__ (filed of the instance, not field of the class)

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+    item_feat_scaler: Optional[MinMaxScaler] = None
+    logger = logging.getLogger("replay")
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:63]: remove from here

Better options right in the module file:
logger = logging.getLogger(__name__)

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 6]
```diff
@@ -0,0 +1,733 @@
...
+from replay.models.distributed_lightfm_wrap.utils import LightFMTraining
+from replay.models.hnswlib import HnswlibMixin
+from replay.session_handler import State
+from replay.utils import (
+    check_numeric,
+    load_pickled_from_parquet,
+    save_picklable_to_parquet,
+    unionify,
+)
+
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:40]: Add logger creation here

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 7]
```diff
@@ -0,0 +1,733 @@
...
+        n_kos_warp: int = 10,
+        max_sampled: int = 100,
+        random_state: Optional[int] = None,
+        tcp_port: int = 21235,
+        connection_timeout: int = 200,
+        hnswlib_params: Optional[dict] = None,
+        num_epochs: int = 30,
+        pygloo_timeout_sec: int = 200,
+    ):
+        if loss == "logistic":
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:83]: add docstring, please

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 8]
```diff
@@ -0,0 +1,733 @@
...
+                ]
+            ),
+        )
+        return (
+            lightfm_factors.join(ids, how="right", on=f"{entity}_idx"),
+            self.model.no_components,
+        )
+
+    @property
+    def _init_args(self):
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:245]: add return type annotation

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 9]
```diff
@@ -0,0 +1,733 @@
...
+            "max_sampled": self.max_sampled,
+            "random_state": self.random_state,
+            "tcp_port": self.tcp_port,
+            "connection_timeout": self.connection_timeout,
+            "hnswlib_params": self._hnswlib_params,
+            "num_epochs": self._num_epochs,
+            "pygloo_timeout_sec": self._pygloo_timeout_sec,
+        }
+
+    def _predict_selected_pairs(
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:264]: add return type annotation

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 10]
```diff
@@ -0,0 +1,733 @@
...
+            raise ValueError("Item features are missing for predict")
+
+        csr_item_features = self._convert_features_to_csr(
+            pairs.select("item_idx").distinct(), item_features
+        )
+        csr_user_features = self._convert_features_to_csr(
+            pairs.select("user_idx").distinct(), user_features
+        )
+
+        return pairs.groupby("user_idx").applyInPandas(
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:293]: groupby seems to be redundant here, having the meaning of the function in  mind.
Probably, can be replaced with direct withColumn("relevance", predict_by_user)

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 11]
```diff
@@ -0,0 +1,733 @@
...
+        self.max_sampled = max_sampled
+        self.random_state = random_state
+        self.tcp_port = tcp_port
+        self.connection_timeout = connection_timeout
+
+        self._hnswlib_params = hnswlib_params
+        self._pygloo_timeout_sec = pygloo_timeout_sec
+        self._num_epochs = num_epochs
+
+        self.model = None
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:105]: I would advise add type annotatons like Optional[...]
For instance, self.num_threads = None -> self.num_threads: Optional[int] = None

Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:105]: Should all these fileds below be public?

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 12]
```diff
@@ -0,0 +1,733 @@
...
+                spark_sess.conf.get("spark.executor.instances", "1")
+            )
+            num_executor_cores = int(
+                spark_sess.conf.get("spark.executor.cores", "1")
+            )
+        self.world_size = max(num_executor_instances, 1)
+        self.num_threads = max(num_executor_cores, 1)
+
+    def _check_parameters(self):
+        if self.loss not in self.losses:
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:342]: replace with assert

assert self.loss in self.losses, f"Only `{'`, `'.join(self.losses)}` losses can be used."

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 13]
```diff
@@ -0,0 +1,733 @@
...
+            raise NotImplementedError(
+                f"Only `{'`, `'.join(self.losses)}` losses can be used."
+            )
+
+    def _initialize_model(self):
+        self.user_feat_scaler = None
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:348]: add a coment that explains why we need to initialize it here

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 14]
```diff
@@ -0,0 +1,733 @@
...
+        self.world_size = max(num_executor_instances, 1)
+        self.num_threads = max(num_executor_cores, 1)
+
+    def _check_parameters(self):
+        if self.loss not in self.losses:
+            raise NotImplementedError(
+                f"Only `{'`, `'.join(self.losses)}` losses can be used."
+            )
+
+    def _initialize_model(self):
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:347]: rename _initialize_model to _reinitialize_model
(_initialize_or_reinitialize_model)

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 15]
```diff
@@ -0,0 +1,733 @@
...
+        returns
+        features: np.float32 csr_matrix of shape [n_users, n_user_features], optional
+             Each row contains that user's weights over features.
+        """
+        if not features:
+            return None
+
+        check_numeric(features)
+
+        entity = "item" if "item_idx" in features.columns else "user"
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:376]: multiple occurences of the same function. Despite it is only a one-liner, still better to move to a separate function

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 16]
```diff
@@ -0,0 +1,733 @@
...
+    ) -> None:
+        self._initialize_model()
+        self._fit_partial(log, user_features, item_features)
+
+    @staticmethod
+    def _get_num_users_and_items(
+        log: DataFrame, previous_log: Optional[DataFrame] = None
+    ) -> Tuple[int, int]:
+        interactions = unionify(log, previous_log)
+        num_users = interactions.agg({"user_idx": "max"}).collect()[0][0]
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:465]: collect()[0][0] - > first()[0]

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 17]
```diff
@@ -0,0 +1,733 @@
...
+        num_items = interactions.agg({"item_idx": "max"}).collect()[0][0]
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:466]: collect()[0][0] - > first()[0]

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 18]
```diff
@@ -0,0 +1,733 @@
...
+    def _reinitialize_embeddings(
+        self,
+        log: DataFrame,
+        previous_log: DataFrame,
+        num_users: int,
+        num_items: int,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+    ) -> None:
+        # Get number of users and items presented in previous log
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:479]: add docstring for input parameters

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 19]
```diff
@@ -0,0 +1,733 @@
...
+        def _concat_arrays(model_array: np.ndarray, added_array: np.ndarray, dissect_at: int):
+            return np.concatenate(
+                (
+                    model_array[:dissect_at],
+                    added_array,
+                    model_array[:dissect_at],
+                )
+            )
+
+        item_updates = [
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:552]: Seems to be the same these two parts, better to get rid of duplicates

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 20]
```diff
@@ -0,0 +1,733 @@
...
+
+        if user_features is not None:
+            self.can_predict_cold_users = True
+        if item_features is not None:
+            self.can_predict_cold_items = True
+
+        log = log.select("user_idx", "item_idx", sf.lit(1).alias("relevance"))
+        interactions = log.repartition(self.world_size, "user_idx")
+
+        n_users, n_items = self._get_num_users_and_items(
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:604]: n_users, n_items = -> self.max_seen_user_idx , self.max_seen_item_idx =

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 21]
```diff
@@ -0,0 +1,733 @@
...
+        log: DataFrame,
+        user_features: Optional[DataFrame] = None,
+        item_features: Optional[DataFrame] = None,
+        previous_log: Optional[DataFrame] = None,
+    ) -> None:
+        if not self.model:
+            self._initialize_model()
+        self._initialize_world_size_and_threads()
+
+        if user_features is not None:
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:596]: if user_features is not None:
    self.can_predict_cold_users = True 
    - >
    
    self.can_predict_cold_users = user_features is not None

Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:596]: if user_features is not None:
    self.can_predict_cold_users = True 
    - >
    
    self.can_predict_cold_users = user_features is not None

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 22]
```diff
@@ -0,0 +1,733 @@
...
+            self.can_predict_cold_users = True
+        if item_features is not None:
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:598]: Same as above

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py [hunk 23]
```diff
@@ -0,0 +1,733 @@
...
+            True,
+            timedelta(seconds=self.connection_timeout),
+        )
+
+        _tcp_port = self.tcp_port
+        _connection_timeout = self.connection_timeout
+        _num_epochs = self._num_epochs
+        _pygloo_timeout = self._pygloo_timeout_sec
+
+        def udf_to_map_on_interactions_with_index(
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/lightfm_wrap.py:670]: Not a comment, but rather TODO for future: need to reconsider whole this function

File: replay/models/distributed_lightfm_wrap/utils.py [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+    user_alpha: np.ndarray
+    item_alpha: np.ndarray
+
+
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+class LightFMTraining:
+    """Trainer used in distributed LightFM."""
+
+    def __init__(self, model: LightFM, world_size: int, num_threads: int):
+        self.model = model
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/utils.py:58]: add docstring

File: replay/models/distributed_lightfm_wrap/utils.py [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+    user_bias_gradients: np.ndarray
+    user_bias_momentum: np.ndarray
+
+    user_alpha: np.ndarray
+    item_alpha: np.ndarray
+
+
+# pylint: disable=too-many-instance-attributes, too-many-arguments
+class LightFMTraining:
+    """Trainer used in distributed LightFM."""
```
Comment (fonhorst) [inline replay/models/distributed_lightfm_wrap/utils.py:55]: add more important details to this docstring

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py (hunk 1 above)
Comment (fonhorst) [reply replay/models/distributed_lightfm_wrap/lightfm_wrap.py:418]: let's leave for now as it is. The logic why you do what you do is clear now.
But add an extensive comments to justify why you do it this way

**Review by fonhorst — CHANGES_REQUESTED** [review]

File: replay/models/distributed_lightfm_wrap/lightfm_wrap.py (hunk 20 above)
Comment (zakharova-anastasiia) [reply replay/models/distributed_lightfm_wrap/lightfm_wrap.py:604]: further, the variables n_users / n_items are used in other functions, to preserve the name<->meaning alignment, variables n_users / n_items are separated from max_seen_user_idx / max_seen_item_idx
